processor/marshalprocessor/processor.go

```diff
@@ -0,0 +1,110 @@
+// Copyright  observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package marshalprocessor
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/zap"
+)
+
+type marshalProcessor struct {
+	logger    *zap.Logger
+	marshalTo string
+}
+
+func newMarshalProcessor(logger *zap.Logger, cfg *Config) *marshalProcessor {
+	return &marshalProcessor{
+		logger:    logger,
+		marshalTo: cfg.MarshalTo,
+	}
+}
+
+// ErrStringBodyNotSupported is returned when a log body is a string
+var ErrStringBodyNotSupported = fmt.Errorf("String body not supported")
+
+func (mp *marshalProcessor) processLogs(_ context.Context, ld plog.Logs) (plog.Logs, error) {
+	var errors []error
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		resourceLog := ld.ResourceLogs().At(i)
+		for j := 0; j < resourceLog.ScopeLogs().Len(); j++ {
+			scopeLog := resourceLog.ScopeLogs().At(j)
+			for k := 0; k < scopeLog.LogRecords().Len(); k++ {
+				logRecord := scopeLog.LogRecords().At(k)
+				logBody := logRecord.Body()
+				// If body is a string, skip that log
+				if logBody.Type() == pcommon.ValueTypeStr {
+					errors = append(errors, ErrStringBodyNotSupported)
+					continue
+				}
+				switch strings.ToUpper(mp.marshalTo) {
+				case "JSON":
+					jsonBody := logBody.AsString()
+					logBody.SetStr(jsonBody)
+				case "XML":
+					return ld, fmt.Errorf("XML not yet supported")
+				case "KV":
+					jsonBody := logBody.AsString()
+					kvBody, err := convertJSONToKV(jsonBody)
+					if err != nil {
+						errors = append(errors, fmt.Errorf("Error converting to KV: %w", err))
+						continue
+					}
+					logBody.SetStr(kvBody)
```

Comment:
This seems inefficient to me. Is there a reason we can't just iterate the key-values of the pcommon.Map instead of marshalling and unmarshalling JSON for every log record?

Reply:
No, that's definitely a better way to do it.